EIPS/eip-erc721-marketplace-extension.md

```diff
@@ -0,0 +1,308 @@
+			saleItems.push( tokenId );
+			unchecked {
+				++itemsForSale;
+			}
+		}
+	}
+
+	function _delistItem( uint256 tokenId ) internal {
+		delete tokenPrice[ tokenId ];
+		delete isForSale[ tokenId ];
+		delete privateRecipient[ tokenId ];
+		unchecked {
+			--itemsForSale;
+		}
+
+		//update token list
+		uint256 tempTokenId = saleItems[ itemsForSale ];
+		if ( tempTokenId == tokenId ) {
+			saleItems.pop();
+		}
+		else {
+			//record the listId of the token we want to get rid of
+			uint256 tempListId = listId[ tokenId ];
+
+			//store the last tokenId from the array into the newly vacant slot
+			saleItems[ tempListId ] = tempTokenId;
+
+			//record the new index of said token
+			listId[ tempTokenId ] = tempListId;
+
+			//remove the last element from the array
+			saleItems.pop();
+		}
+	}
+
+	function _beforeTokenTransfer( address from, address to, uint256 tokenId ) internal override {
+		super._beforeTokenTransfer( from, to, tokenId );
+
+		// if the token is still marked as listed then we need to delist it
+		if ( isForSale[ tokenId ] ) {
+			_delistItem( tokenId );
+		}
+
+		emit Transfer( from, to, tokenId );
+	}
+
+	// PUBLIC
+	function listItem( uint256 tokenId, uint256 price, address to ) external {
+		address tokenOwner = ownerOf( tokenId );
+		require( msg.sender == tokenOwner, "Invalid Lister" );
+		_listItem( tokenId, price );
+		privateRecipient[ tokenId ] = to;
+		emit Listed( tokenId, tokenOwner, to, price );
+	}
+
+	function delistItem( uint256 tokenId ) external {
+		require( isForSale[ tokenId ], "Item not for Sale" );
+		require( msg.sender == ownerOf( tokenId ), "Invalid Lister" );
+		_delistItem( tokenId );
+		emit Delisted( tokenId );
+	}
+
+	function buyItem( uint256 tokenId ) external payable {
+		require( isForSale[ tokenId ], "Item not for Sale" );
+		uint256 totalPrice = tokenPrice[ tokenId ];
+		require(msg.value == totalPrice, "Incorrect price");
+		( address royaltyRecipient, uint256 royaltyPrice ) = royaltyInfo( tokenId, totalPrice );
+
+		address buyer = msg.sender;
+		if ( privateRecipient[ tokenId ] != address( 0 ) ) {
+			require( buyer == privateRecipient[ tokenId ], "invalid sale address" );
+		}
+
+		address tokenOwner = ownerOf( tokenId );
+		_delistItem( tokenId );
+		emit Purchased( tokenId, tokenOwner, buyer, totalPrice );
+
+		( bool success, ) = payable( tokenOwner ).call{ value: totalPrice - royaltyPrice }( "" );
+		require( success, "Transaction Unsuccessful" );
+
+		if (royalty > 0){
+			// Can also be set to payout directly to specified wallet
+			// alternately this block can be removed to save gas and 
+			// royalties will be stored on the smart contract.
+			( success, ) = payable( royaltyRecipient ).call{ value: royaltyPrice }( "" );
+			require( success, "Transaction Unsuccessful" );
+		}
+
+		_transfer( tokenOwner, buyer, tokenId );
+	}
+
+	// VIEW
+	function getAllListings() external view returns ( Listing[] memory ) {
+		uint256 arraylen = saleItems.length;
```

Comment:
Wouldn't it be better to have pagination? What if the marketplace was very large?

```solidity
function getListings(uint startIndex, uint fetchCount) external view returns(Listing[] memory) {
  uint itemCount = salesItems.length;
  if(itemCount == 0) {
    return new Listing[](0);
  }
  // Error if starting after end
  require(startIndex < itemCount);
  // Trim if fetchCount goes beyond the upper bound
  if(startIndex + fetchCount >= itemCount) {
    fetchCount = itemCount - startIndex;
  }
  Listing[] memory out = new Listing[](fetchCount);
  for(uint i; i < fetchCount; i++) {
    out[i] = salesItems[startIndex + i];
  }
  return out;
}
```

Reply:
Except it's not accessing data from "another contract." That would defeat the entire purpose of the code. You might be right that putting some kind of index basis on it could be helpful, as could adding some kind of fetch or get that limits by the cost of the listings or any other criteria you see fit. So perhaps the better solution would be to remove getAllListings from the EIP and leave it up to the developer to create sorting algorithms.
The other way would be to include such searches in here to make it more universal for front-end searches so that aggregators will have a better interfact to gather data from a wide number of compliant collections.

I will look into salesItems again. There was some reason it was left out and for brevity sake we wanted to provide a solution to customers immediately who did not wish to comply with OpenSeas blacklist functionality but still wanted to have control of their own nfts on their own marketplace.  Will look into that more this morning to see about cleaning things up.

Reply:
I've updated salesItems array like you mentioned. There were some other conflicts when I initially wrote the code but they have resolved themselves now. There still exists two mappings to track indexes for individual tokens to avoid the need to loop through the whole array, and the return values are now much cleaner for the current two fetch functions.
Your method of fetching several at a time should also be feasible, but I have not implemented it yet.

I've tasked @lambdalf-dev with updating the code here on the EIP, it should be updated shortly. Also includes removal of a bug that I found while retesting the codebase.